### PR TITLE
feat(curriculum): create callback curriculum

### LINKED ISSUE #55052: Dedicated section teaching callbacks
### Is your feature request related to a problem? Please describe.

Campers do not understand callbacks well.

I see this everywhere on the forum. Even with people far into the backend part of the curriculum. It happens with higher order functions, with handler functions, with callbacks such as done, take your pick.

I know we technically teach them as part of teaching other subjects but, we might want to consider a short section dedicated just to using callbacks, just as we have one on async.

### Describe the solution you'd like

A new short section dedicated to callbacks.

### Describe alternatives you've considered

It might be possible to combine it with teaching async. But there isn't really a one to one correlation in all cases.

### Additional context

Edit: Maybe we also make it a point to teach what is meant by functions as first-class citizens/First-class Function

https://developer.mozilla.org/en-US/docs/Glossary/First-class_Function

Again, to expand. The main issue is they are not able to identify a callback and do not understand how to use a callback, so they tend to invoke the function or not understand the idea of passing a function around as a value.

### PR DESCRIPTION
**Pull request by a2937**
Checklist:

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #55052

I have began proper work on the Callback project. I wish to ensure that I am on the right track so far.

### CODE REVIEW

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675b238e128702f011e5395.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+title: Step 1
+challengeType: 1
+dashedName: step-1
+---
+
+# --description--
+
+Callbacks are among the most powerful tools in Javascript. Callbacks are used to pass a function into another function. The passed in function is then executed inside of the other function. The output then influences the other function.
+
+Let's start building out an example. Begin by creating a function called `sayHello` that returns "Hi".
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675b238e128702f011e5395.md:12]:
```suggestion
Let's start building out an example. Begin by creating a function called `sayHello` that returns `"Hi"`. 
```

Strings need to be in code tags when they refer to the user's code.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675b238e128702f011e5395.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+# --hints--
+
+You need to define `sayHello` as a function. 
+
+```js
+assert.isFunction(sayHello); 
+```
+
+You need to make `sayHello` return "Hi," with no extra characters.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675b238e128702f011e5395.md:22]:
```suggestion
You need to make `sayHello` return `"Hi"`, with no extra characters. 
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675df027b8c9574fa7bbffa.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
+---
+id: 6675df027b8c9574fa7bbffa
+title: Step 2
+challengeType: 1
+dashedName: step-2
+---
+
+# --description--
+
+Now that we have created a function that returns "Hi", let's create a second function to pass into
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675df027b8c9574fa7bbffa.md:10]:
```suggestion
Now that we have created a function that returns `"Hi"`, let's create a second function to pass into 
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675df027b8c9574fa7bbffb.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+challengeType: 1
+dashedName: step-3
+---
+
+# --description--
+
+For this lesson, we are going to make it possible to actually pass a function as a variable
+by adding an parameter to our `sayHello` function called `retrievedName`. 
+
+The `retrievedName` variable can also be referred to as a `First-class function` as it a function
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675df027b8c9574fa7bbffb.md:13]: It's not, yet.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675df027b8c9574fa7bbffc.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+---
+id: 6675df027b8c9574fa7bbffc
+title: Step 4
+challengeType: 1
+dashedName: step-4
+---
+
+# --description--
+
+Now we are going to actually use the function we passed in earlier. Modify the return statement of `sayHello` to include
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-callbacks-by-saying-hello/6675df027b8c9574fa7bbffc.md:10]: This is where I'd mention first class function

**Review by naomi-lgbt — COMMENTED** [review]
Just an early peek

**Comment by lasjorg** [comment]
All functions in JS are first class functions. The term merely refers to the ability to treat functions as values. It isn't related to the specific use of the function. I would mention it in the beginning as an overall language construct when explaining what a callback is.

---

Not sure what the plan is for this challenge right now. But I was thinking about this the other day, and maybe writing a tiny testing library (like a small part of Jest) might be one option.

They have well-defined use of callbacks, and it might be a topic worth briefly exploring all on its own.

**Comment by a2937** [comment]
How should I incorporate Jest into this? I didn't realize a browser only version of it existed?

**Comment by lasjorg** [comment]
We wouldn't incorporate Jest, we would build a tiny testing lib that works like Jest (or Mocha, or whatever).

```
test('adds 1 + 2 to equal 3', () => {
  expect(sum(1, 2)).toBe(3);
});
```

Example code:

`https://github.com/kentcdodds/js-testing-fundamentals/blob/main/lessons/testing-framework.js`

---

I would also suggest some examples where the same higher-order function is passed different callbacks and produce different results. Like `map` and a couple of callbacks that transforms the array elements in various ways.

**Comment by a2937** [comment]
I am going to reopen a new PR since practically everything in it will change.

**Pull request closed without merging**